### PR TITLE
date provider: using a ConstSingleton date formatter

### PR DESCRIPTION
**Pull request by alyssawilk**
We've seen shutdown crashes in E-M due to the deletion of this static object.

Risk Level: low
Testing: internal tsan now green
Docs Changes: n/a
Release Notes: n/a

Fixes https://github.com/envoyproxy/envoy/issues/26091

### CODE REVIEW

**Review by RyanTheOptimist — DISMISSED** [review]
Thanks for doing this. Would it make sense to consider creating a NonConstSingleton class for this purpose? Alternatively, perhaps a brief comment about how this "leaks" but we're fine with it?

**Review by RyanTheOptimist — APPROVED** [review]

File: source/common/http/date_provider_impl.cc [hunk 1]
```diff
@@ -6,7 +6,12 @@
 namespace Envoy {
 namespace Http {
 
-DateFormatter DateProviderImplBase::date_formatter_("%a, %d %b %Y %H:%M:%S GMT");
+class TlsCachingDateFormatter : public DateFormatter {
```
Comment (RyanTheOptimist) [inline source/common/http/date_provider_impl.cc:9]: What is TLS specific about this class? (I'm guessing that means Thread Local Storage not Transport Level Security, but I don't think I see what is thread local about this?

nit: Maybe this should be in an anon namespace?

File: source/common/http/date_provider_impl.cc [hunk 2]
```diff
@@ -6,7 +6,12 @@
...
+public:
+  TlsCachingDateFormatter() : DateFormatter("%a, %d %b %Y %H:%M:%S GMT") {}
+};
+
+using GlobalTlsCachingDateFormatter = ConstSingleton<TlsCachingDateFormatter>;
```
Comment (RyanTheOptimist) [inline source/common/http/date_provider_impl.cc:14]: Ditto about TLS. Maybe `DateFormatterSingleton`?

**Review by snowp — APPROVED** [review]

**Review by RyanTheOptimist — COMMENTED** [review]
Ah, nice!

File: source/common/http/date_provider_impl.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/date_provider_impl.cc]: done

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**